STM32F405RGTx/Core/Src/rtos_threads/state_machine.cpp

```diff
@@ -0,0 +1,460 @@
+    SetLedColour(0.0, 50.0, 0.0);
+/*     State_t normal_fault_check = NormalFaultChecking();
+    State_t severe_fault_check = SevereFaultChecking();
+    if (severe_fault_check == SevereDangerFault) {
+        return severe_fault_check;
+    }
+    else if (normal_fault_check == NormalDangerFault) {
+        return normal_fault_check;
+    }
+ */
+    // TODO: Make list of the threads that need to be turned on or turned off
+        // On: Measurements Thread, CANZThread
+        // Off: PIDThread, VHzThread, SVPWMThread
+    MeasurementsThread::resumeMeasurements();
+
+    if (!Queue_empty(&RX_QUEUE)) {
+        CANFrame rx_frame = CANBus_get_frame();
+        uint8_t state_id = CANFrame_get_field(&rx_frame, STATE_ID);
+        idle_state_id = state_id;
+        if (state_id == AUTO_PILOT) {
+            return AutoPilot;
+        }
+        else if (state_id == MANUAL_OPERATION_WAITING) {
+            return ManualControl;
+        } 
+    }
+
+    return Idle;
+}
+
+State_t StateMachineThread::AutoPilotEvent(void)
+{
+    // Set LED colour to yellow
+    SetLedColour(50.0, 50.0, 0.0);
+
+    // Send ACK on CAN when stop complete 
+    if (NewState != CurrentState) {
+        CANFrame tx_frame = CANFrame_init(MOTOR_CONTROLLER_STATE_CHANGE_ACK_NACK);
+        CANFrame_set_field(&tx_frame, MOTOR_CONTROLLER_STATE_ID_ACK_NACK, idle_state_id);
+        if (CANBus_put_frame(&tx_frame) != HAL_OK) { Error_Handler(); }
+    }
+
+    // TODO: Make list of the threads that need to be turned on or turned off
+        // On: Measurements Thread, CANThread
+        // Off: PIDThread, VHzThread, SVPWMThread
+    MeasurementsThread::resumeMeasurements();
+
+    // Receive CAN frame
+    if (!Queue_empty(&RX_QUEUE)) {
+        CANFrame rx_frame = CANBus_get_frame();
+        uint8_t state_id = CANFrame_get_field(&rx_frame, STATE_ID);
+        int distance_to_end = TRACK_LENGTH - g_mc_data.curr_pos;
+        if (state_id == EMERGENCY_BRAKE || state_id == SYSTEM_FAILURE) {
+            return SevereDangerFault;
+            // TODO: Talk to Ryan/Dev when to enter MinorDangerFault
+        }
+        else if (state_id == BRAKING || distance_to_end <= DISTANCE_THRESHOLD) {
+            return Idle;
+        } 
+    }
+    return AutoPilot;
+}
+
+State_t StateMachineThread::ManualControlEvent(void)
+{
+    // Set LED colour to blue
+    SetLedColour(0, 0, 50.0);
+    // Send ACK on CAN when stop complete
+    if (NewState != CurrentState) {
+        CANFrame tx_frame = CANFrame_init(MOTOR_CONTROLLER_STATE_CHANGE_ACK_NACK);
+        CANFrame_set_field(&tx_frame, MOTOR_CONTROLLER_STATE_ID_ACK_NACK, idle_state_id);
+        if (CANBus_put_frame(&tx_frame) != HAL_OK) { Error_Handler(); }
+
+        CANFrame tx_frame1 = CANFrame_init(MANUAL_CONTROL_1);
+        uint8_t target_speed = CANFrame_get_field(&tx_frame1, TARGET_SPEED);
+        CANFrame_set_field(&tx_frame1, TARGET_SPEED, target_speed);
+        // Not sure what to put for target frequency
```

Comment:
You're not supposed to be transmitting a manual control frame, the MC just receives these to determine these. You are currently making an empty frame with no payload and trying to extract the target speed from it and then setting the frame's field again with the same number you just got